[Issue-39] Python improvement

gaol opened the pull request:
Motivation: Fixes #39 

Explain here the context, and why you're making that change, what is the problem you're trying to solve.

Conformance:

Your commits should be signed and you should have signed the Eclipse Contributor Agreement as explained in https://github.com/eclipse/vert.x/blob/master/CONTRIBUTING.md
Please also make sure you adhere to the code style guidelines: https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines


This PR tries to improve Python client of eventbus client by:
* renamed to follow PEP conventions
* removed unnecessary DelveryOption, it can be defined as headers on sending messages to bridge
* added integration tests via project https://github.com/gaol/test-eventbus-bridge
* re connect(5 times by default) if current connection is broken
* introduce logging module for the logging messages
* REAEME update

punkstarman commented on python/vertx/eventbus.py:
```diff
@@ -0,0 +1,346 @@
+            try:
+                if self._state != _State.CONNECTED:
+                    self._state = _State.CONNECTING
+                    self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
Why omit the last two characters of `socket`?

punkstarman commented on python/vertx/eventbus.py:
```diff
@@ -0,0 +1,346 @@
+                    self.sock.settimeout(self.timeout)
+                    self.sock.connect((self.host, self.port))
+                    self._state = _State.CONNECTED
+                    # receiving thread
+                    t1 = Thread(target=self._receive)
```
The variable name is too short. With a better name, the comment becomes unnecessary.

punkstarman commented on python/vertx/eventbus.py:
```diff
@@ -0,0 +1,346 @@
+                    t1 = Thread(target=self._receive)
+                    t1.setDaemon(True)
+                    t1.start()
+                    tp = Thread(target=self._ping)
```
The variable name is too short.

punkstarman commented on python/vertx/eventbus.py:
```diff
@@ -0,0 +1,346 @@
+        while self.is_connected():
+            try:
+                LOGGER.debug("send ping")
+                # check last pong
+                if self.last_pong is not None:
+                    if time.time() - self.last_pong > self.ping_interval * 2:
+                        LOGGER.warning("ping/pong packet is slow")
```
Why not write a method for this called `check_last_pong`?

punkstarman commented on python/vertx/eventbus.py:
```diff
@@ -0,0 +1,346 @@
+
+    def send(self, address, headers=None, body=None, reply_address=None, reply_handler=None):
+        self._check_closed()
+        ra = reply_address
+        rh = reply_handler
```
These variable names are too short. Why not just use the original method parameters and modify `reply_address` when it is `None`?
While modifying parameters can make code difficult to read, if it's just to handle a default value, it's OK.

punkstarman commented on python/vertx/eventbus.py:
```diff
@@ -0,0 +1,346 @@
+        ra = reply_address
+        rh = reply_handler
+        if rh is not None:
+            if ra is None:
+                ra = str(uuid.uuid1())
```
```suggestion
            reply_address = reply_address or str(uuid.uuid1())
```

punkstarman reviewed (changes requested):
This sort of PR is much needed because the current code is not Pythonic. The code in this PR is a step in the right direction. It isn't entirely PEP8 compliant, so a spot of flake8 or maybe even [Black](https://black.readthedocs.io/en/stable/) would be useful. The tests work nicely :+1: 

I wonder if the refactoring isn't a bit too ambitious. Since both the tests and the implementation were refactored, it's difficult to judge whether there are any regressions. Splitting this into two separate steps might have made reviewing easier.

gaol commented on python/README.md:
```diff
@@ -5,41 +5,48 @@ This is a TCP eventbus implementation for python clients.
+ebus = EventBus('localhost', 7000)
 
-eb=Vertx.Eventbus(Client(),'localhost', 7000)
+# Define a handler to handle messgae
```
@punkstarman Thank you for the review :)

> Splitting this into two separate steps might have made reviewing easier.

I will check to see how to split it for easier review.

gaol commented on python/vertx/eventbus.py:
```diff
@@ -0,0 +1,346 @@
+                            for handler in self.handlers[message['address']].all_handlers():
+                                handler(message)
+                        else:
+                            LOGGER.warning("No handler found on address %s" % message['address'])
```
should the `%s` be removed as well if it is changed to `,` ?

punkstarman replied:
No. Logging methods provide formatting functionality. See [logging.Logger.debug](https://docs.python.org/3/library/logging.html#logging.Logger.debug).

gaol commented on python/vertx/eventbus.py:
```diff
@@ -0,0 +1,346 @@
+    BROKEN = 5  # when the client connection is broken
+
+
+class EventBus:
```
:+1: make sense :)

gaol commented on python/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+vertx-eventbus-bridge-starter
```
> Why not start by including it directly in this project, and then separating it when needed?

The intention of the test project is not only for Python client, but also for other clients. You are right that it makes no sense to have a test only dependency declared in the `requirements.txt`.

punkstarman replied:
> The intention of the test project is not only for Python client, but also for other clients.

I totally understand the intention, but the quality isn't up to speed, and it isn't because the first use case here isn't mature. This is normal. Including the test project here for the first step seems safer to me, and it doesn't preclude it from being separated later when it is more mature and ready to be shared by other projects.

gaol replied:
> it doesn't preclude it from being separated later when it is more mature and ready to be shared by other projects.

OK, I think it makes sense, thank you :)  The test project includes both Java code of the bridge server, and the Python code on how to start the bridge, I will move the Python one only to the test in this project, and leave the Java one as where it is to make this project clean. 

> `requirements.txt` is meant for development dependencies. Runtime dependencies are often specified using `setuptools` in a `setup.py` file

I see, thank you :) This `vertx-eventbus-client` should not depend on non-standard libraries. I will remove the `requirements.txt` file because of including the test project directly.

gaol commented on python/vertx/eventbus.py:
```diff
@@ -0,0 +1,346 @@
+        frame = struct.pack('!I', len(message)) + message
+        self.sock.sendall(frame)
+
+    def send(self, address, headers=None, body=None, reply_address=None, reply_handler=None):
```
:+1:  the ability of checking `time out` of when the `reply_handler` gets called was missing. thanks :)

gaol commented:
@punkstarman I updated according to your feedback, thank you very much on the reviews. 

The update includes:
* review feedback on README and logging messages
* make the EventBus context manager
* format the code using [Black](https://pypi.org/project/black/)
* leave the test codes unchanged, focusing on implementation itself. current test codes actually test nothing IMO :)
* checking reply timeout when specifying reply_handler on send method
* removed uneeded DeliveryOption

The update does not include:
* test codes against new implementation

Hope it can make the review easier, :) once this gets approved, I will send another PR to improve the tests. :)

gaol commented:
Thank you @punkstarman for the review. I am merging this and start with the tests part :)
